### PR TITLE
Added an option to enable a password check against a Banned-Password List

### PR DESCRIPTION
**Pull request by 2403905**
## Description

Added an option to enable a password check against a Banned-Password List

## Related Issue

- Fixes [https://jira.owncloud.com/browse/OCIS-3809](urhttps://jira.owncloud.com/browse/OCIS-3809)

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [X] Code changes
- [X] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: services/frontend/README.md [hunk 1]
```diff
@@ -74,13 +74,17 @@ Generally, a password can contain any UTF-8 characters, however some characters
...
--   `FRONTEND_PASSWORD_POLICIES_MIN_UPPERCASE_CHARACTERS`
+-   `FRONTEND_PASSWORD_POLICY_MIN_UPPERCASE_CHARACTERS`
 Define the minimum number of lowercase letters.
--   `FRONTEND_PASSWORD_POLICIES_MIN_DIGITS`
+-   `FRONTEND_PASSWORD_POLICY_MIN_DIGITS`
 Define the minimum number of digits.
--   `FRONTEND_PASSWORD_POLICIES_MIN_SPECIAL_CHARACTERS`
+-   `FRONTEND_PASSWORD_POLICY_MIN_SPECIAL_CHARACTERS`
 Define the minimum number of special characters.
+-  `FRONTEND_BANNED_PASSWORDS_LIST`
```
Comment (mmattel) [inline services/frontend/README.md:89]:
```suggestion
-   `FRONTEND_BANNED_PASSWORDS_LIST`
```

File: services/frontend/pkg/config/config.go [hunk 2]
```diff
@@ -176,9 +176,10 @@ type ServiceAccount struct {
...
+	MinCharacters          int    `yaml:"min_characters,omitempty" env:"FRONTEND_PASSWORD_POLICY_MIN_CHARACTERS" desc:"Define the minimum password length. Defaults to 0 if not set."`
+	MinLowerCaseCharacters int    `yaml:"min_lowercase_characters" env:"FRONTEND_PASSWORD_POLICY_MIN_LOWERCASE_CHARACTERS" desc:"Define the minimum number of uppercase letters. Defaults to 0 if not set."`
+	MinUpperCaseCharacters int    `yaml:"min_uppercase_characters" env:"FRONTEND_PASSWORD_POLICY_MIN_UPPERCASE_CHARACTERS" desc:"Define the minimum number of lowercase letters. Defaults to 0 if not set."`
+	MinDigits              int    `yaml:"min_digits" env:"FRONTEND_PASSWORD_POLICY_MIN_DIGITS" desc:"Define the minimum number of digits. Defaults to 0 if not set."`
+	MinSpecialCharacters   int    `yaml:"min_special_characters" env:"FRONTEND_PASSWORD_POLICY_MIN_SPECIAL_CHARACTERS" desc:"Define the minimum number of characters from the special characters list to be present. Defaults to 0 if not set."`
+	BannedPasswordsList    string `yaml:"banned_passwords_list" env:"FRONTEND_BANNED_PASSWORDS_LIST" desc:"The path to the banned passwords list file."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:184]:
```suggestion
	BannedPasswordsList    string `yaml:"banned_passwords_list" env:"FRONTEND_BANNED_PASSWORDS_LIST" desc:"Path to the 'banned passwords list' file."`
```

File: changelog/unreleased/add-passwod-banned-password-list.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+Enhancement: Add the Banned-Passwords List
+
+Added an option to enable a password check against a Banned-Password-List OCIS-3809
```
[the remark this answers is not in the document]
Comment (EParzefall) [reply changelog/unreleased/add-passwod-banned-password-list.md:3]: No dashes in English.

File: services/frontend/pkg/config/config.go [hunk 4]
```diff
@@ -176,9 +176,10 @@ type ServiceAccount struct {
...
+	MinCharacters          int    `yaml:"min_characters,omitempty" env:"FRONTEND_PASSWORD_POLICY_MIN_CHARACTERS" desc:"Define the minimum password length. Defaults to 0 if not set."`
+	MinLowerCaseCharacters int    `yaml:"min_lowercase_characters" env:"FRONTEND_PASSWORD_POLICY_MIN_LOWERCASE_CHARACTERS" desc:"Define the minimum number of uppercase letters. Defaults to 0 if not set."`
+	MinUpperCaseCharacters int    `yaml:"min_uppercase_characters" env:"FRONTEND_PASSWORD_POLICY_MIN_UPPERCASE_CHARACTERS" desc:"Define the minimum number of lowercase letters. Defaults to 0 if not set."`
+	MinDigits              int    `yaml:"min_digits" env:"FRONTEND_PASSWORD_POLICY_MIN_DIGITS" desc:"Define the minimum number of digits. Defaults to 0 if not set."`
+	MinSpecialCharacters   int    `yaml:"min_special_characters" env:"FRONTEND_PASSWORD_POLICY_MIN_SPECIAL_CHARACTERS" desc:"Define the minimum number of characters from the special characters list to be present. Defaults to 0 if not set."`
+	BannedPasswordsList    string `yaml:"banned_passwords_list" env:"FRONTEND_PASSWORD_POLICY_BANNED_PASSWORDS_LIST" desc:"Path to the 'banned passwords list' file."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:184]:
```suggestion
	BannedPasswordsList    string `yaml:"banned_passwords_list" env:"FRONTEND_PASSWORD_POLICY_BANNED_PASSWORDS_LIST" desc:"Path to the 'banned passwords list' file. See the documentation for more details."`
```
I have seen the readme text part too late, just adding the standard reference text 😄

File: changelog/unreleased/add-passwod-banned-password-list.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (2403905) [reply changelog/unreleased/add-passwod-banned-password-list.md:3]: Thanks. Just copy/paste from a ticket

**Review by mmattel — APPROVED** [review]
LGTM from a docs-pov 👍

File: services/frontend/pkg/revaconfig/config.go [hunk 5]
```diff
@@ -301,3 +317,34 @@ func FrontendConfigFromStruct(cfg *config.Config) (map[string]interface{}, error
 		},
 	}, nil
 }
+
+func readMultilineFile(path string) (interface{}, error) {
```
Comment (kobergj) [inline services/frontend/pkg/revaconfig/config.go:321]: Please return a `map[string]struct{}`. Or is there a reason for returning an interface?

File: services/frontend/pkg/revaconfig/config.go [hunk 6]
```diff
@@ -301,3 +317,34 @@ func FrontendConfigFromStruct(cfg *config.Config) (map[string]interface{}, error
...
+		}
+	}
+	return data, err
+}
+
+func fileExists(path string) bool {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
```
Comment (kobergj) [inline services/frontend/pkg/revaconfig/config.go:345]: Please also return false `if err != nil` just to be on the safe side

**Review by kobergj — CHANGES_REQUESTED** [review]

File: services/frontend/pkg/revaconfig/config.go (hunk 6 above)
Comment (2403905) [reply services/frontend/pkg/revaconfig/config.go:345]: If we return false when error it doesn't make sense to check  `os.IsNotExist(err)` before.

File: services/frontend/pkg/revaconfig/config.go [hunk 7]
```diff
@@ -1,24 +1,39 @@
...
 	webURL.Path = path.Join(webURL.Path, "external")
 	webOpenInAppURL := webURL.String()
 
+	var bannedPasswordsList map[string]struct{}
+	if cfg.PasswordPolicy.BannedPasswordsList != "" {
+		bannedPasswordsList, err = readMultilineFile(cfg.PasswordPolicy.BannedPasswordsList)
+		if err != nil {
+			logger.Err(err).
+				Str("server", cfg.Service.Name).
+				Msgf("Failed to load the banned passwords from a file %s, skipped.", cfg.PasswordPolicy.BannedPasswordsList)
```
Comment (kobergj) [inline services/frontend/pkg/revaconfig/config.go:33]: We should return the error here. If someone wants to set a banned password list, they must immediately know that it didn't work.

**Review by kobergj — CHANGES_REQUESTED** [review]
Just one point. Rest looks fine. Will approve after reva is adjusted

**Review by kobergj — APPROVED** [review]

**Pull request merged**